contracts/ScheduledPayment.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.9;
+
+import "@gnosis.pm/zodiac/contracts/core/Module.sol";
+import "@openzeppelin/contracts-upgradeable/utils/structs/EnumerableSetUpgradeable.sol";
+
+contract ScheduledPayment is Module {
+    using EnumerableSetUpgradeable for EnumerableSetUpgradeable.Bytes32Set;
+
+    event ScheduledPaymentSetup(
+        address indexed initiator,
+        address indexed owner,
+        address indexed avatar,
+        address target,
+        address config
+    );
+    event PaymentScheduled(uint256 nonce, bytes32 spHash);
+
+    event ConfigSet(address config);
+
+    address public config;
+    uint256 public nonce;
+    EnumerableSetUpgradeable.Bytes32Set private spHash;
+
+    modifier onlyAvatar() {
+        require(msg.sender == avatar, "caller is not the right avatar");
+        _;
+    }
+
+    constructor(
+        address _owner,
+        address _avatar,
+        address _target,
+        address _config
+    ) {
+        bytes memory initParams = abi.encode(_owner, _avatar, _target, _config);
+        setUp(initParams);
+    }
+
+    function setUp(bytes memory initParams) public override initializer {
+        (
+            address _owner,
+            address _avatar,
+            address _target,
+            address _config
+        ) = abi.decode(initParams, (address, address, address, address));
+        __Ownable_init();
+        require(_avatar != address(0), "Avatar can not be zero address");
+        require(_target != address(0), "Target can not be zero address");
+
+        avatar = _avatar;
+        target = _target;
+        config = _config;
+
+        transferOwnership(_owner);
+
+        emit ScheduledPaymentSetup(
+            msg.sender,
+            _owner,
+            _avatar,
+            _target,
+            _config
+        );
+    }
+
+    function schedulePayment(bytes32 _spHash) external onlyAvatar {
```

Comment:
I just want to confirm that the reason we don't hash on chain is because we don't want to leak the details of the scheduled payment correct?

Reply:
if that is the case, it might not be a bad idea to include in comments what the hash is composed of.

Reply:
I think so, @alex-cardstack could you help to confirm?

Reply:
Confirmed.

Maybe it might be good to have a `pure` function that can be called in web3 so that the hashing _code_ is solidity but it can be executed by local simulation?

A comment is less work but not executable. I don't have a strong feeling either way.

Reply:
not a bad idea--that way there is only ever a single implementation that is on and off chain, which could prevent issues if the hashing function needed to change but we didn't update both onchain and offchain implementations in lockstep

Reply:
I've added the function to create SP hash